neuraxle/base.py

```diff
@@ -1740,33 +1741,33 @@ def should_resume(self, data_container: DataContainer, context: ExecutionContext
             return True
         return False
 
-    def apply(self, method_name: str, *kargs, **kwargs) -> 'BaseStep':
+    def apply(self, method_name: str, *kargs, **kwargs) -> Union[Dict, Iterable]:
         """
         Apply the method name to the meta step and its wrapped step.
 
         :param method_name: method name that need to be called on all steps
         :param kargs: any additional arguments to be passed to the method
         :param kwargs: any additional positional arguments to be passed to the method
-        :return: self (not a new step)
-        :rtype: BaseStep
+        :return: accumulated results
+        :rtype: Union[Dict, Iterable]
         """
-        BaseStep.apply(self, method_name, *kargs, **kwargs)
-        self.wrapped.apply(method_name, *kargs, **kwargs)
-        return self
+        results = BaseStep.apply(self, method_name, *kargs, **kwargs)
+        wrapped_results = self.wrapped.apply(method_name, *kargs, **kwargs)
+        return join_apply_results(results, wrapped_results)
 
-    def apply_method(self, method: Callable, *kargs, **kwargs) -> 'BaseStep':
+    def apply_method(self, method: Callable, *kargs, **kwargs) -> Union[Dict, Iterable]:
         """
         Apply method to the meta step and its wrapped step.
 
         :param method: method to call with self
         :param kargs: any additional arguments to be passed to the method
         :param kwargs: any additional positional arguments to be passed to the method
-        :return: self (not a new step)
-        :rtype: BaseStep
+        :return: accumulated results
+        :rtype: Union[Dict, Iterable]
         """
-        BaseStep.apply_method(self, method, *kargs, **kwargs)
-        self.wrapped = self.wrapped.apply_method(method, *kargs, **kwargs)
-        return self
+        results = BaseStep.apply_method(self, method, *kargs, **kwargs)
+        wrapped_results = self.wrapped.apply_method(method, *kargs, **kwargs)
+        return join_apply_results(results, wrapped_results)
```

Comment:
I'd probably prefer using a dict like this: 

```
return {
    "SKLearnWrapper": results,
    "SKLearnWrapper_wrapped": wrapped_results
}
```

Reply:
Otherwise the `join_apply_results` must be a method passed in argument.

Reply:
mmmm okey applying your changes again, but I don't think this will make this method easy to use.

Reply:
@guillaume-chevalier hey btw it is the method that is being called by apply that returns the keys for the dictionary so I think this code is fine.

Reply:
@guillaume-chevalier you would want to pass the names of the each pipeline steps to the apply methods ?